### PR TITLE
Sub-daily Time Constraints

### PR DESCRIPTION
**Pull request by courtneyholcomb**
Enable sub-daily time constraints.

### CODE REVIEW

File: metricflow-semantics/metricflow_semantics/time/dateutil_adjuster.py [hunk 1]
```diff
@@ -66,18 +66,35 @@ def expand_time_constraint_to_fill_granularity(
...
+        """Takes a date and adjusts it to the start of the given granularity period.
+
+        Used to adjust time constraints when the time dimension they will be applied to has a larger granularity
+        than the time constraint requested.
+
+        Args:
+            time_granularity: The granularity of the time dimension that will be constrained.
+            date_to_adjust: The time constraint start date.
+        """
+        date_to_adjust = deepcopy(date_to_adjust)
```
Comment (plypaul) [inline metricflow-semantics/metricflow_semantics/time/dateutil_adjuster.py:78]: Nit: `datetime` objects are immutable, so you don't need to copy them.

File: metricflow-semantics/metricflow_semantics/time/dateutil_adjuster.py [hunk 2]
```diff
@@ -100,18 +117,35 @@ def adjust_to_start_of_period(
...
+        """Takes a date and adjusts it to the end of the given granularity period.
+
+        Used to adjust time constraints when the time dimension they will be applied to has a larger granularity
+        than the time constraint requested.
+
+        Args:
+            time_granularity: The granularity of the time dimension that will be constrained.
+            date_to_adjust: The time constraint end date.
+        """
+        date_to_adjust = deepcopy(date_to_adjust)
```
Comment (plypaul) [inline metricflow-semantics/metricflow_semantics/time/dateutil_adjuster.py:129]: Ditto.

**Review by plypaul — COMMENTED** [review]
Is it useful to have seconds in [test_start_and_end_periods__results.txt](https://github.com/dbt-labs/metricflow/pull/1369/files#diff-4f1d5c4e77d0c0d4d09053a431c3a6f31946f1117e010fa40dc5f5f4b6831e7b)? The file size is very large.

**Comment by courtneyholcomb** [comment]
> Is it useful to have seconds in [test_start_and_end_periods__results.txt](https://github.com/dbt-labs/metricflow/pull/1369/files#diff-4f1d5c4e77d0c0d4d09053a431c3a6f31946f1117e010fa40dc5f5f4b6831e7b)? The file size is very large.

@plypaul Added a commit to remove `SECOND`. It should reduce the file size by 1/6th, but it's only iterating through days (not seconds) so not sure if that's as dramatic of a change as you expected.

**Review by plypaul — APPROVED** [review]
Actually, how about we just keep it to day to keep the snapshot the same? That test was mainly to check peculiarities with weeks / quarters / years.

**Pull request merged**